src/huggingface_hub/commit_api.py

```diff
@@ -0,0 +1,444 @@
+import base64
+import io
+import os
+from dataclasses import dataclass
+from functools import partial
+from hashlib import sha256
+from math import ceil
+from os.path import getsize
+from typing import Dict, Generator, Iterable, List, Literal, Optional, TypedDict, Union
+
+import requests
+from requests.auth import HTTPBasicAuth
+
+from .constants import ENDPOINT, REPO_TYPES_URL_PREFIXES
+from .utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+
+UploadMode = Literal["lfs", "regular"]
+
+
+@dataclass
+class CommitOperationDelete:
+    path_in_repo: str
+
+
+@dataclass
+class CommitOperationAdd:
+    path_in_repo: str
+    path_or_fileobj: Union[str, bytes, io.BufferedIOBase]
+
+    def __post_init__(self):
```

Comment:
validation should happen when we need to use the data. A user can construct a class with some data, and then change it later in which case `__post_init__` won't be triggered.